File: frontend/src/types/prediction.ts
export interface ShapContribution {
  feature: string
  feature_value: number
  shap_value: number
  direction: 'increases_emission' | 'decreases_emission'
}

export interface PredictionExplanation {
  base_value: number
  prediction: number
  top_drivers: ShapContribution[]
  explanation_method: string
}

export interface ForecastResult {
  company_id: number
  scope: string
  category: string
  reporting_year: number
  reporting_month: number
  predicted_co2_tonnes: number
  explanation: PredictionExplanation
}

export interface AnomalyRecord {
  record_id: number
  company_id: number
  scope: string
  category: string
  co2_tonnes: number
  reporting_year: number
  reporting_month: number
  anomaly_score: number
  is_anomaly: boolean
  anomaly_severity: 'high' | 'medium' | 'low' | null
}

export interface AnomalyResult {
  company_id: number
  year: number
  total_records: number
  anomaly_count: number
  anomaly_rate: number
  records: AnomalyRecord[]
}

export interface FeatureImportanceItem {
  feature: string
  importance: number
  rank: number
}

export interface ModelStatus {
  forecasting_model: boolean
  anomaly_detector: boolean
}

File: app/schemas/prediction.py
from pydantic import BaseModel, Field
from typing import Optional
from app.models.emission import EmissionScope, EmissionCategory


class ForecastRequest(BaseModel):
    """Request body for POST /predictions/forecast."""
    company_id: int = Field(..., gt=0)
    scope: EmissionScope
    category: EmissionCategory
    reporting_year: int = Field(..., ge=2000, le=2100)
    reporting_month: int = Field(..., ge=1, le=12)


class ShapContribution(BaseModel):
    feature: str
    feature_value: float
    shap_value: float
    direction: str


class PredictionExplanation(BaseModel):
    base_value: float
    prediction: float
    top_drivers: list[ShapContribution]
    explanation_method: str


class ForecastResponse(BaseModel):
    company_id: int
    scope: str
    category: str
    reporting_year: int
    reporting_month: int
    predicted_co2_tonnes: float
    explanation: PredictionExplanation


class AnomalyScanRequest(BaseModel):
    """Request body for POST /predictions/anomalies."""
    company_id: int = Field(..., gt=0)
    year: int = Field(..., ge=2000, le=2100)


class AnomalyRecordResult(BaseModel):
    record_id: int
    company_id: int
    scope: str
    category: str
    co2_tonnes: float
    reporting_year: int
    reporting_month: int
    anomaly_score: float
    is_anomaly: bool
    anomaly_severity: Optional[str] = None


class AnomalyScanResponse(BaseModel):
    company_id: int
    year: int
    total_records: int
    anomaly_count: int
    anomaly_rate: float
    records: list[AnomalyRecordResult]


class FeatureImportanceItem(BaseModel):
    feature: str
    importance: float
    rank: int


class FeatureImportanceResponse(BaseModel):
    feature_importance: list[FeatureImportanceItem]


class ModelStatusResponse(BaseModel):
    forecasting_model: bool
    anomaly_detector: bool
